### PR TITLE
feat(metanode): for mp leader send snapshot to follower:

### LINKED ISSUE #2149: [Bug]: meta partition raft follower may apply wrong cursor when apply snapshot from leader
**Describe the bug**
each meta partition  maintains a cursor for generate inodeId. the cursor can only be monotonically incremented.
which means onece inodeId=N is generated, inodeIds be generated later must be greate than N, even the inodeId=N has been deleted.

But when meta partition raft leader send snapshot to followers, the cursor is not contained in snapshot.
when applies snapshot, the follower use the max inodeId of snop.Op opFSMCreateInode as the cursor.
But  the  max inodeId of snop.Op opFSMCreateInode may be less than the actual cursor of leader! see this senario:
(1)  meta partition's current max inodeID of files in the FS is N, cursor is N.
(2)  the file of inodeID=N is deleted, max inodeID of files in the FS becomes N-M, cursor is still N
(3)  the leader took an raft local snapshot, and the raft log of deletion of inodeID=N has been truncated out.
(4) leader send snapshot to follwer,  the follwer took inodeID=N-M as cursor, but leader's cursor is still N.

**Environment**
Describe your server system version, kernel version
CubeFS version < 3.3.0

**To Reproduce**
(1) modify func (m *MetaNode) startRaftServe(), set NumOfLogsToRetain as 1, to make reproduce more easyly.
(2) rebuild the codes, deploy a cluster, create a volume and mount it.
(3) create some files in the mount point path, choose a meta partition to test, and make it has mutiple inodes.
(4) delete the file of max inodeID of the chosen meta partition.
(5) make meta partition deommission onf the chosen meta partition, trigger the leader send snapshot to the newly add-in follower.
(6) check the curors of leader and th newly add-in follow, the former is larger than the latter

**Expected behavior**
when follower's cursor should be consistent with the leader when  apply snapshot.

**Additional context**
Solution suggestion：
meta partition's raft leader sends snapshot to follower,  add cursor filed in the snapshot, to make follower's apply cursor consistent with the leader.

Compatibility consideration：
In cluster upgrading senario, metaNodes of different snapshot versions exists at the same time,
to avoid problems, snapshot is not allowed to be applied if snapshot versions not match.
Need to think carefully about compatibility issues of upgrading.

### PR DESCRIPTION
**Pull request by true1064**
**What this PR does / why we need it**:
for mp leader send snapshot to follower:
(1) add cursor in snapshot;
(2) when apply snapshot, return error if snapshot format not match;

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #2149

**Special notes for your reviewer**:
return error if snapshot format not match when apply snapshot, so need to think carefully about compatibility issues of upgrading.

### CODE REVIEW

File: metanode/partition_fsm.go [hunk 1]
```diff
@@ -487,23 +488,49 @@ func (mp *metaPartition) ApplySnapshot(peers []raftproto.Peer, iter raftproto.Sn
...
+		// Consider the following 2 cases:
+		// (1) leader's version is after transaction feature, follower's is before transaction feature:
+		//     the follower will return an error in the switch-block below, for lower version will consider snap.Op == opFSMCursorSnapshot
+		//     as an unknown op.
+		// (2) If leader's version is before transaction feature, follower's is after transaction feature:
+		//     follower will detect the snapshot format it not match in the if-block below
+		if index == 2 && snap.Op != opFSMCursorSnapshot {
+			log.LogErrorf("ApplySnapshot: error for snapshot format not match, partitionID(%v), index:%v, expect snap.Op:%v, actual snap.Op:%v",
+				mp.config.PartitionId, index, opFSMCursorSnapshot, snap.Op)
+			err = fmt.Errorf("snapshot format not match")
```
Comment (xiangcai1215) [inline metanode/partition_fsm.go:523]: 这里用err=errors.New("snapshot format not match")

Comment (true1064) [reply metanode/partition_fsm.go:523]: 这也用有也行吧？switch的default block原本用的就是err = fmt.Errorf("unknown op=%d", snap.Op)

**Review by xiangcai1215 — DISMISSED** [review]

File: metanode/partition_item.go [hunk 2]
```diff
@@ -148,12 +150,45 @@ type MetaItemIterator struct {
 	closeOnce sync.Once
 }
 
+// snap wrapper  key definition
+const (
+	swKeyCursor uint32 = iota
```
[the remark this answers is not in the document]
Comment (true1064) [reply metanode/partition_item.go:155]: swKeyCursor is for program extensibility.  the data structure can also be used by new fields in the future,  distinguished by the key field

**Review by JasonHu520 — APPROVED** [review]

**Review by xiangcai1215 — APPROVED** [review]

File: metanode/partition_item.go [hunk 3]
```diff
@@ -148,22 +156,46 @@ type MetaItemIterator struct {
...
+	return
+}
+
 // newMetaItemIterator returns a new MetaItemIterator.
 func newMetaItemIterator(mp *metaPartition) (si *MetaItemIterator, err error) {
 	si = new(MetaItemIterator)
 	si.fileRootDir = mp.config.RootDir
 	si.applyID = mp.applyID
 	si.txId = mp.txProcessor.txManager.txIdAlloc.getTransactionID()
+	si.cursor = mp.config.Cursor
```
[the remark this answers is not in the document]
Comment (true1064) [reply metanode/partition_item.go]: done

**Pull request merged**